lib/syscall_shim/include/uk/syscall.h

```diff
@@ -294,6 +326,168 @@ typedef long uk_syscall_arg_t;
+		__UK_SYSCALL_RETADDR_ENTRY();				\
+		errno = 0;						\
+		ret = ename(						\
+			UK_ARG_MAPx(x, UK_S_ARG_CAST_LONG, __VA_ARGS__)); \
+		__UK_SYSCALL_RETADDR_CLEAR();				\
+		if (ret == -1)						\
+			ret = errno ? -errno : -EFAULT;			\
+		errno = _errno;						\
+		return ret;						\
+	}								\
+	static inline rtype __##ename(UK_ARG_MAPx(x,			\
+					UK_S_ARG_ACTUAL, __VA_ARGS__)); \
+	long ename(UK_ARG_MAPx(x, UK_S_ARG_LONG, __VA_ARGS__))		\
+	{								\
+		long ret;						\
+									\
+		__UK_SYSCALL_PRINTD(x, rtype, ename, __VA_ARGS__);	\
+		__UK_SYSCALL_RETADDR_ENTRY();				\
+		ret = (long) __##ename(					\
+			UK_ARG_MAPx(x, UK_S_ARG_CAST_ACTUAL, __VA_ARGS__)); \
+		__UK_SYSCALL_RETADDR_CLEAR();				\
+		return ret;						\
+	}								\
+	static inline rtype __##ename(UK_ARG_MAPx(x,			\
+						  UK_S_ARG_ACTUAL_MAYBE_UNUSED,\
+						  __VA_ARGS__))
+#define _UK_LLSYSCALL_DEFINE2(...) __UK_LLSYSCALL_DEFINE2(__VA_ARGS__)
+#define UK_LLSYSCALL_DEFINE2(rtype, name, ...)				\
+	_UK_LLSYSCALL_DEFINE2(__UK_SYSCALL_DEF_NARGS(__VA_ARGS__),	\
+			      rtype,					\
+			      name,					\
+			      __UK_NAME2SCALLE_FN(name),		\
+			      __UK_NAME2SCALLR_FN(name),		\
+			      __VA_ARGS__)
+
+/*
+ * UK_SYSCALL_DEFINE2()
+ * Based on UK_LLSYSCALL_DEFINE and provides a libc-style wrapper
+ * in case UK_LIBC_SYSCALLS is enabled
+ */
+#if UK_LIBC_SYSCALLS
+#define __UK_SYSCALL_DEFINE2(x, rtype, name, ename, rname, ...)		\
+	long ename(UK_ARG_MAPx(x, UK_S_ARG_LONG, __VA_ARGS__));		\
+	rtype name(UK_ARG_MAPx(x, UK_S_ARG_ACTUAL, __VA_ARGS__))	\
+	{								\
+		return (rtype) ename(					\
+			UK_ARG_MAPx(x, UK_S_ARG_CAST_LONG, __VA_ARGS__)); \
+	}								\
+	__UK_LLSYSCALL_DEFINE2(x, rtype, name, ename, rname, __VA_ARGS__)
+#define _UK_SYSCALL_DEFINE2(...) __UK_SYSCALL_DEFINE2(__VA_ARGS__)
+#define UK_SYSCALL_DEFINE2(rtype, name, ...)				\
+	_UK_SYSCALL_DEFINE2(__UK_SYSCALL_DEF_NARGS(__VA_ARGS__),		\
+			    rtype,					\
+			    name,					\
+			    __UK_NAME2SCALLE_FN(name),			\
+			    __UK_NAME2SCALLR_FN(name),			\
+			    __VA_ARGS__)
+#else
+#define UK_SYSCALL_DEFINE2(rtype, name, ...)				\
+	_UK_LLSYSCALL_DEFINE2(__UK_SYSCALL_DEF_NARGS(__VA_ARGS__),	\
+			      rtype,					\
+			      name,					\
+			      __UK_NAME2SCALLE_FN(name),		\
+			      __UK_NAME2SCALLR_FN(name),		\
+			      __VA_ARGS__)
+#endif /* UK_LIBC_SYSCALLS */
+
+/* Raw system call implementation that is returning negative codes on errors */
+/* TODO: `void` as return type is currently not supported.
+ * NOTE: Workaround is to use `int` instead.
+ */
+/*
+ * UK_LLSYSCALL_R_DEFINE2()
+ * Low-level variant, does not provide a libc-style wrapper
+ */
+#define __UK_LLSYSCALL_R_DEFINE2(x, rtype, name, ename, rname, ...)	\
+	long rname(UK_ARG_MAPx(x, UK_S_ARG_LONG, __VA_ARGS__));		\
+	long ename(UK_ARG_MAPx(x, UK_S_ARG_LONG, __VA_ARGS__))		\
+	{								\
+		long ret;						\
+									\
+		ret = rname(UK_ARG_MAPx(x, UK_S_ARG_CAST_LONG, __VA_ARGS__)); \
+		__UK_SYSCALL_RETADDR_ENTRY();				\
```

Comment:
Shouldn't they be reversed?

Reply:
Thanks a lot for seeing this. Truely a mistake. The code is updated and changed now.